### PR TITLE
Issue 180

### PR DESCRIPTION
**Pull request by GulderBone**

### CODE REVIEW

File: src/test/java/com/cognifide/aemrules/matcher/MethodMatcherTest.java [hunk 1]
```diff
@@ -19,26 +19,18 @@
...
-import org.sonar.java.resolve.SemanticModel;
-import org.sonar.plugins.java.api.tree.ClassTree;
-import org.sonar.plugins.java.api.tree.CompilationUnitTree;
-import org.sonar.plugins.java.api.tree.ExpressionStatementTree;
-import org.sonar.plugins.java.api.tree.MethodInvocationTree;
-import org.sonar.plugins.java.api.tree.MethodTree;
-import org.sonar.plugins.java.api.tree.StatementTree;
-import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.java.model.JParser;
+import org.sonar.plugins.java.api.tree.*;
```
Comment (jplucinski) [inline src/test/java/com/cognifide/aemrules/matcher/MethodMatcherTest.java]: Please avoid multi imports.

File: src/test/java/com/cognifide/aemrules/matcher/MethodMatcherTest.java [hunk 2]
```diff
@@ -148,12 +140,16 @@ public void shouldNotMatchMethodWhenMethodParameterTypesDoNotMatch() {
...
-        SemanticModel.createFor(compilationUnitTree, new SquidClassLoader(Collections.emptyList()));
+        CompilationUnitTree compilationUnitTree = parse(codeToParse);
         ClassTree classTree = (ClassTree) compilationUnitTree.types().get(CLASS_INDEX);
         StatementTree statementTree = ((MethodTree) classTree.members().get(CLASS_METHOD_INDEX)).block().body().get(METHOD_INVOCATION_INDEX);
         this.methodInvocationTree = (MethodInvocationTree) ((ExpressionStatementTree) statementTree).expression();
     }
 
+    private CompilationUnitTree parse(String source) {
+        List<File> classpath = Lists.newArrayList(new File("target/test-classes"), new File("target/classes"));
+        return JParser.parse("1.8", "test", source, classpath);
```
Comment (jplucinski) [inline src/test/java/com/cognifide/aemrules/matcher/MethodMatcherTest.java]: Please extract string into static constant variables.

File: src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java [hunk 3]
```diff
@@ -88,8 +89,8 @@ public void visitAssignmentExpression(AssignmentExpressionTree tree) {
   @Override
   public void visitTryStatement(TryStatementTree tree) {
     insideTryStatement = true;
-    tree.resources().stream()
-        .map(resource -> resource.simpleName().name())
+    tree.resourceList().stream()
+        .map(resource -> ((VariableTreeImpl) resource).simpleName().name())
```
Comment (toniedzwiedz) [inline src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java]: What's the reason behind referring to an implementation type here and casting?

**Comment by toniedzwiedz** [comment]
SonarCloud claims there's no test coverage on two lines you changed here. Also, it seems to still be using @chutch 's personal SonarCloud account.

File: src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java [hunk 4]
```diff
@@ -88,8 +89,8 @@ public void visitAssignmentExpression(AssignmentExpressionTree tree) {
   @Override
   public void visitTryStatement(TryStatementTree tree) {
     insideTryStatement = true;
-    tree.resources().stream()
-        .map(resource -> resource.simpleName().name())
+    tree.resourceList().stream()
```
Comment (toniedzwiedz) [inline src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java]: Are you sure this is synonymous to the previous line? SonarCloud seems to think these lines aren't covered by unit tests. This means that they never were or that existing tests stopped following this execution path. Please investigate.

Comment (jplucinski) [reply src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java]: In my opinion issue here is related to travis config. Covervall can measure coverage. I think that this issue should be moved to new one as switching to other sonar cloud account.

File: src/test/java/com/cognifide/aemrules/matcher/MethodMatcherTest.java (hunk 2 above)
Comment (GulderBone) [reply src/test/java/com/cognifide/aemrules/matcher/MethodMatcherTest.java]: done

File: src/test/java/com/cognifide/aemrules/matcher/MethodMatcherTest.java (hunk 1 above)
Comment (GulderBone) [reply src/test/java/com/cognifide/aemrules/matcher/MethodMatcherTest.java]: changed

File: src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java (hunk 3 above)
Comment (GulderBone) [reply src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java]: This was the only way I found to get that prop.

File: src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java (hunk 4 above)
Comment (jplucinski) [reply src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java]: There was missing configuration for Jacoco reports, it's fixed now.

Comment (jplucinski) [reply src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java]: Issue for build config:
https://github.com/Cognifide/AEM-Rules-for-SonarQube/issues/183

File: src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java (hunk 3 above)
Comment (toniedzwiedz) [reply src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java]: `tree.resourceList()` returns a `ListTree<Tree>` and the method you're trying to use here doesn't exist at such a high interface level. Still, you don't need to use the specific implementation type here. As far as I can tell, the interface `VariableTree` should be sufficient

```
         // We're only interested in variables so filter out VariableTree instances from a list of Tree. Not all trees have names.
        .filter(resource -> resource instanceof VariableTree)
        .map(VariableTree.class::cast)
        .map(resource -> resource.simpleName().name())
        .forEach(resourceResolversInTryWithResources::add);
```

I'm trying to think of a unit test case that would show us where this cast could fail without the filter.

Comment (toniedzwiedz) [reply src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java]: I can't think of a way to pass a syntax tree that doesn't cast to `VariableTree` inside a `try` statement's resource list that is parse-able as valid Java, as per https://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.20.3

 I wonder why the `resourceTree` method, which appears specific to the `TryStatementTree` doesn't return something more specific than `Tree`. Could be just a matter of consistency.

How about:
```
tree.resourceList().stream()
  // We're iterating over a resource specification in a try-with-resource block
  // so we expect variable trees
  .filter(resource -> resource instanceof VariableTree)
  .map(VariableTree.class::cast)
  .map(resource -> resource.simpleName().name())
  .forEach(resourceResolversInTryWithResources::add);
```

Comment (toniedzwiedz) [reply src/main/java/com/cognifide/aemrules/java/checks/resourceresolver/close/ResourceResolverTryWithResourcesCheck.java]: This kind of works but it turns out there's a new way to write a try-with-resources block in Java 9 and above.
https://stackoverflow.com/questions/61124346/when-does-sonars-trystatementtreeresourcelist-return-a-tree-that-isnt-a-varia/61124506#61124506

Issue https://github.com/Cognifide/AEM-Rules-for-SonarQube/issues/188 raised to address this.

**Pull request merged**